file_writer_control/CommandChannel.py

```diff
@@ -0,0 +1,217 @@
+
+from file_writer_control.InThreadStatusTracker import InThreadStatusTracker
+from file_writer_control.WorkerStatus import WorkerStatus, WorkerState
+
+from file_writer_control.JobStatus import JobStatus, JobState
+from file_writer_control.CommandStatus import CommandStatus, CommandState
+from file_writer_control.InThreadStatusTracker import (
+    STATUS_MESSAGE_TIMEOUT,
+    COMMAND_STATUS_TIMEOUT,
+    JOB_STATUS_TIMEOUT,
+)
+
+
+def thread_function(host_port: str, topic: str, in_queue: Queue, out_queue: Queue):
+    """
+    Background thread for consuming Kafka messages.
+    :param host_port: The host + port of the Kafka broker that we are using.
+    :param topic: The Kafka topic that we are listening to.
+    :param in_queue: A queue for sending "exit" messages to the thread.
+    .. note:: The queue will exit upon the reception of the string "exit" on this queue.
+    :param out_queue: The queue to which status updates are published.
+    """
+    status_tracker = InThreadStatusTracker(out_queue)
+    while True:
+        try:
+            consumer = KafkaConsumer(
+                topic,
+                bootstrap_servers=host_port,
+                fetch_max_bytes=52428800 * 6,
+                consumer_timeout_ms=100,
+            )  # Roughly 300MB
+            break
+        except NoBrokersAvailable:
+            pass
+        if not in_queue.empty():
+            new_msg = in_queue.get()
+            if new_msg == "exit":
+                return
+    while True:
+        for message in consumer:
+            status_tracker.process_message(message.value)
+        status_tracker.check_for_lost_connections()
+        if not in_queue.empty():
+            new_msg = in_queue.get()
+            if new_msg == "exit":
+                break
+    consumer.close(True)
+
+
+class CommandChannel(object):
+    """
+    A class that implements the functionality for receiving and interpreting messages that are published to the
+    Kafka command topic of a pool of file-writers.
+    """
+
+    def __init__(self, command_topic_url: str):
+        """
+        Constructor.
+        :param command_topic_url: The url of the Kafka topic to where the file-writer status/command messages are published.
+        """
+        kafka_address = KafkaTopicUrl(command_topic_url)
+        self.status_queue = Queue()
+        self.to_thread_queue = Queue()
+        thread_kwargs = {
+            "host_port": kafka_address.host_port,
+            "topic": kafka_address.topic,
+            "in_queue": self.to_thread_queue,
+            "out_queue": self.status_queue,
+        }
+        self.map_of_workers: Dict[str, WorkerStatus] = {}
+        self.map_of_jobs: Dict[str, JobStatus] = {}
+        self.map_of_commands: Dict[str, CommandStatus] = {}
+        self.run_thread = True
+        self.thread = threading.Thread(
+            target=thread_function, daemon=True, kwargs=thread_kwargs
+        )
+        self.thread.start()
+
+        def do_exit():
```

Comment:
Why this is wrapped up by `__init__`? Is there a specific technical reason?

Reply:
Yes, and the reason is one of the weaknesses of Python. Note that the thread has `daemon=True`. This is the only way to make the thread automatically stop on exit. However, we do not want to kill the thread without a proper shutdown and thus leaving the Kafka broker waiting for a reply that will never come from the consumer that we have created. Doing it like this was the only way I could figure out how to make the thread automatically stop in a (somewhat) graceful fashion.